Fix #12365: Company inauguration date should use calendar year in wallclock mode

Vinixss opened the pull request:
<!--
Commit message:

- Please use Feature / Add / Change / Fix for player-facing changes. E.g.: "Feature: My cool new feature".
- Please use Feature / Add / Change / Fix followed by "[NewGRF]" or "[Script]" for moddable changes. E.g.: "Feature: [NewGRF] My cool new NewGRF addition".
- Please use Codechange / Codefix for developer-facing changes. E.g.: "Codefix #1234: Validate against nullptr properly".

See https://github.com/OpenTTD/OpenTTD/blob/master/CODINGSTYLE.md#commit-message for more details.
-->

## Motivation / Problem
When using the Wallclock timekeeping setting, the Inauguration year of the company is displayed in minutes rather than the actual calendar year

Closes #12365. (edit 2TT)
<!--
Describe here shortly
* For bug fixes:
    * What problem does this solve?
    * If there is already an issue, link the issue, otherwise describe the problem here.
* For features or gameplay changes:
    * What was the motivation to develop this feature?
    * Does this address any problem with the gameplay or interface?
    * Which group of players do you think would enjoy this feature?
-->


## Description
Added a new variable to company_base.h named "inaugurated_year_calendar" to store the calendar time version of the inaugurated year. Both this and the economic inaugurated_year are passed as parameters into the company window in company_gui.cpp in order to display both.
<!--
Describe here shortly
* For bug fixes:
    * How is the problem solved?
* For features or gameplay changes:
    * What does this feature do?
    * How does it improve/solve the situation described under 'motivation'.
-->


## Limitations
Main limitation is that, in order to fix the bug in the way described under Issue #12365, I had to edit english.txt and add a new value, so other translation files do not have this
<!--
Describe here
* Is the problem solved in all scenarios?
* Is this feature complete? Are there things that could be added in the future?
* Are there things that are intentionally left out?
* Do you know of a bug or corner case that does not work?
-->


## Checklist for review
Some things are not automated, and forgotten often. This list is a reminder for the reviewers.
* This PR touched english.txt

rubidium42 committed:
Fix: ensure revision mismatch also considers random debugging status

PeterN committed:
Codechange: Store both left and right glyph positions in a run.

This also allows the end of a run to be known without using an extra position entry.

PeterN committed:
Codechange: Return read-only span instead of vector from layout run functions.

PeterN committed:
Codechange: Move GetCharPosInString/GetCharAtPosition to gfx_layout.

These functions are related more to layouting than graphics.

PeterN committed:
Change: Position caret on left or right of glyph depending on language direction.

steve-goldman committed:
Change: Do not automatically connect double depots with track

JGRennison committed:
Fix #12030: Water infrastructure totals when using DC_FORCE_CLEAR_TILE

To remove objects on water

Vinixss committed:
Added Calendar verion of the inauguration year

Commit:
Update: Translations from eints
vietnamese: 3 changes by KhoiCanDev
polish: 13 changes by pAter-exe

PeterN committed:
Fix #12648: Ensure all uses of std::filesystem::path use native encoding. (#12650)

std::filesystem::path does not expect UTF-8 strings, so they must be converted to native format first (and back to utf-8 after.)

Vinixss committed:
Fix: Added back SetDParam() for inauguration_year

Vinixss committed:
Fix #12365: Added second parameter for INAUGURATED_TITLE to show both calendar and economic inaugurated years

James103 commented on src/lang/english.txt:
```diff
@@ -3886,7 +3886,7 @@ STR_FINANCES_INFRASTRUCTURE_BUTTON                              :{BLACK}Infrastr
 STR_COMPANY_VIEW_PRESIDENT_MANAGER_TITLE                        :{WHITE}{PRESIDENT_NAME}{}{GOLD}(Manager)
 
-STR_COMPANY_VIEW_INAUGURATED_TITLE                              :{GOLD}Inaugurated: {WHITE}{NUM}
+STR_COMPANY_VIEW_INAUGURATED_TITLE                              :{GOLD}Inaugurated: {WHITE}{NUM} (period {NUM})
```
This should be split into two strings: one containing just `{GOLD}Inaugurated: {WHITE}{NUM}` for calendar timekeeping, and the new string `{GOLD}Inaugurated: {WHITE}{NUM} (period {NUM})` for wallclock timekeeping.

Vinixss replied:
Thank you, will do that

Commit:
Update: Translations from eints
galician: 3 changes by pvillaverde
danish: 1 change by bscargo
polish: 5 changes by pAter-exe

2TallTyler commented on src/company_base.h:
```diff
@@ -92,6 +92,7 @@ struct CompanyProperties {
 	TileIndex last_build_coordinate; ///< Coordinate of the last build thing by this company.
 
 	TimerGameEconomy::Year inaugurated_year; ///< Economy year of starting the company.
+    TimerGameCalendar::Year inaugurated_year_calendar; ///< Calendar year of starting the company. Used for when we need to display an actual Inaguration Year while in Wallclock mode.
```
```suggestion
    TimerGameCalendar::Year inaugurated_year_calendar; ///< Calendar year of starting the company. Used for when we need to display an actual Inauguration Year while in Wallclock mode.
```

2TallTyler reviewed (commented):
Thank you for the contribution! You will need to clean up your commit history with an [interactive rebase](https://thoughtbot.com/blog/git-interactive-rebase-squash-amend-rewriting-history#interactive-rebase) and then force-push. Feel free to ask if you have any questions!

PeterN committed:
Codechange: Pass WindowDesc by reference instead of pointer. (#12771)

WindowDesc as passed to Windows is not optional so don't allow to it to be nullptr.

2TallTyler committed:
Doc: Remove resolved issue #4415 from known-bugs.txt (#12773)

Commit:
Update: Translations from eints
german: 62 changes by Wuzzy2

PeterN committed:
Codechange: Pass NWidgetParts as span instead of begin/end pointers. (#12779)

Commit:
Update: Translations from eints
polish: 1 change by pAter-exe

PeterN committed:
Codechange: Shuffle OrderList members to reduce size.

Size is reduced from 40 bytes to 32 bytes on 64 bit platforms.

PeterN committed:
Codechange: Shuffle RoadStop members to reduce size.

Size is reduced from 40 bytes to 32 bytes on 64 bit platforms.

PeterN committed:
Fix #12775: Text truncated incorrectly on Windows. (#12782)

Caused by a -1 in the wrong place, which moved glyphs left one pixel.

PeterN committed:
Codechange: Use std::endian instead of TTD_ENDIAN where trivial. (#12778)

PeterN committed:
Codechange: Pass NewGRF ByteReader by reference. (#12786)

ByteReader must be present and not null, so don't use a pointer.

PeterN committed:
Codechange: Split MakeNWidget to improve readability. (#12785)

Split MakeNWidget() into two stages, widget-creation and attribute-applying, to reduce function size and make it clearer.

glx22 committed:
Fix #12787: boostrap doesn't have main vindow (#12788)

2TallTyler committed:
Fix #10239: Use elevated snow height for roads on flat foundations (#12776)

2TallTyler committed:
Change: Don't use house construction states in Scenario Editor (#12730)

ladysadie committed:
Doc: Explain comment alignment in CODINGSTYLE.md

merni-ns committed:
Add: Setting to disable warning for old vehicles (#12714)

Commit:
Update: Translations from eints
greek: 2 changes by gh658804
lithuanian: 1 change by khamper

ladysadie committed:
Codechange: Compiler check for _fios_colours/DetailedFileType (#12781)

PeterN committed:
Codefix: Incorrect indentation of NWidgetParts in load/save windows. (#12791)

Fix to use the "standard" layout of 1 tab for nesting widgets and 2 tabs for attributes that wrap on to the next line.

2TallTyler committed:
Codefix: Don't claim to allocate new waypoint when just checking (#12792)

Vinixss committed:
Revert "Added SetDParam for economic Inauguration Year back in so both economic and calendar inauguration years can be visible"
-m "Made mistake in editing translation files"

This reverts commit 95c9770dc6b1532f0e392931684ac10c59f7f7d2.

Commit:
Update: Translations from eints
english (au): 2 changes by krysclarke
norwegian (bokmal): 6 changes by eriksorngard
korean: 3 changes by telk5093
russian: 2 changes by Ln-Wolf
finnish: 3 changes by hpiirai
portuguese: 2 changes by azulcosta
portuguese (brazilian): 2 changes by pasantoro
polish: 5 changes by pAter-exe

PeterN committed:
Codefix: Add missing `const` inside script functions. (#12794)

Scripts do not modify items directly, marking them const enforces this.

2TallTyler commented on src/company_gui.cpp:
```diff
@@ -2400,9 +2404,13 @@ struct CompanyWindow : Window
+                    SetDParam(0, Company::Get((CompanyID) this->window_number)->inaugurated_year_calendar);
+                    SetDParam(1, Company::Get((CompanyID) this->window_number)->inaugurated_year);
+                }
+                else SetDParam(0, Company::Get((CompanyID) this->window_number)->inaugurated_year);
```
`else` goes on the same line as the closing bracket for the `if`, and also needs brackets. See my string parameters snippet above for the correct syntax.

PeterN committed:
Codechange: Avoid making copies of intermediate layout runs. (#12796)

The vector of runs is not used after it is passed to the ParagraphLayout class, so pass with std::move to avoid an unnecessary copy.

2TallTyler commented on src/company_base.h:
```diff
@@ -92,6 +92,7 @@ struct CompanyProperties {
 	TileIndex last_build_coordinate; ///< Coordinate of the last build thing by this company.
 
 	TimerGameEconomy::Year inaugurated_year; ///< Economy year of starting the company.
+    TimerGameCalendar::Year inaugurated_year_calendar; ///< Calendar year of starting the company. Used for when we need to display an actual Inauguration Year while in Wallclock mode.
```
We use tabs for indentation, not spaces.

2TallTyler commented on src/company_gui.cpp:
```diff
@@ -2237,6 +2237,10 @@ struct CompanyWindow : Window
 			}
 		}
 
+        if (TimerGameEconomy::UsingWallclockUnits()) {
+            this->GetWidget<NWidgetLeaf>(WID_C_DESC_INAUGURATION)->SetDataTip(STR_COMPANY_VIEW_INAUGURATED_TITLE_WALLCLOCK, STR_NULL);
+        }
+
```
Good try, but we typically use a different way to change strings in GUI.

First, in the widget structure use a string that allows us to select strings as a parameter. In this case, you need to set two string parameters, so it should be STR_JUST_STRING2, like so:

```
NWidget(WWT_TEXT, COLOUR_GREY, WID_C_DESC_INAUGURATION), SetDataTip(STR_JUST_STRING2, STR_NULL), SetFill(1, 0),
```

Then choose the correct string in SetStringParameters:

```
case WID_C_DESC_INAUGURATION:
    if (TimerGameEconomy::UsingWallclockUnits()) {
		SetDParam(0, STR_COMPANY_VIEW_INAUGURATED_TITLE_WALLCLOCK);
		SetDParam(1, Company::Get((CompanyID) this->window_number)->inaugurated_year_calendar);
        SetDParam(2, Company::Get((CompanyID) this->window_number)->inaugurated_year);
	} else {
		SetDParam(0, STR_COMPANY_VIEW_INAUGURATED_TITLE);
		SetDParam(1, Company::Get((CompanyID)this->window_number)->inaugurated_year);
	}
    break;
```

2TallTyler commented on src/company_cmd.cpp:
```diff
@@ -608,6 +608,7 @@ Company *DoStartupNewCompany(bool is_ai, CompanyID company = INVALID_COMPANY)
 	c->avail_railtypes = GetCompanyRailTypes(c->index);
 	c->avail_roadtypes = GetCompanyRoadTypes(c->index);
 	c->inaugurated_year = TimerGameEconomy::year;
+    c->inaugurated_year_calendar = TimerGameCalendar::year;
```
Intent with tabs, not spaces.

2TallTyler reviewed (commented):
You will need to add savegame conversion, so when somebody loads a savegame created before this data was saved, something will show up. I'm not sure what to show here, but here's how you'd do it: First, create a new SaveLoadVersion entry in `saveload.h`, like so:

```
SLV_COMPANY_INAUGURATED_PERIOD,         ///< 337  PR#12769 Companies show the period inaugurated in wallclock mode.
```
Then add the conversion code in `afterload.cpp`, like so:
```
if (IsSavegameVersionBefore(SLV_COMPANY_INAUGURATED_PERIOD)) {
	for (Company *c : Company::Iterate()) {
		c->inaugurated_year_calendar = (whatever value you want to set)
	}
}
```
You will need to squash all your commits into one in order for CI to succeed, otherwise the commit checker will fail.

Also, stop clicking the `Update branch` button on GitHub: it adds merge commits and makes a mess of your PR. As long as the check says `Merging can be performed automatically with 1 approving review.`, you do not need to update anything. 😃
